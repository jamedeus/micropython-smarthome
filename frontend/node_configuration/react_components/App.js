import React from 'react';
import InstanceCard from './InstanceCard';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import './Animation.css';


// Takes object and key prefix, returns all keys that begin with prefix
function filterObject(obj, prefix) {
    return Object.entries(obj).reduce((acc, [key, value]) => {
        if (key.startsWith(prefix)) {
            acc[key] = value;
        }
        return acc;
    }, {});
};


class App extends React.Component {
    constructor(props) {
        super(props);
        // Default state object if not received from context
        this.state = {
            metadata: {},
            wifi: {}
        };
    }

    componentDidMount() {
        // Overwrite state with config received from context (if present)
        const config = JSON.parse(document.getElementById("config").textContent);
        console.log(config)
        if (config) {
            this.setState({ ...config });
        }
    }

    logState = () => {
        console.log(this.state);
    };

    // Handler for add device and add sensor buttons
    addInstance = (category) => {
        this.setState(prevState => {
            // Get index of new instance
            const index = Object.keys(filterObject(prevState, category)).length + 1;

            // Add key to state object with empty config
            // Will be populated with metadata template when user selects type
            prevState[`${category}${index}`] = {}

            return prevState;
        });
    };

    // Handler for delete button on device and sensor cards
    deleteInstance = (id) => {
        this.setState(prevState => {
            // Remove deleted instance from state object
            delete prevState[id]
            // Decrement all subsequent IDs in state object to prevent gaps in index
            return update_ids(id, prevState);
        });
    };

    // Handler for type select dropdown in device and sensor cards
    changeInstanceType = (id, category, event) => {
        this.setState(prevState => {
            // Get config template for selected type from metadata object
            const template = metadata[`${category}s`][event.target.value]['config_template'];
            // Overwrite section in state object with config template
            prevState[id] = template;
            return prevState;
        })
    };

    // Handler for all inputs inside device and sensor cards
    // Takes device/sensor ID, param, and value; updates state object
    // TODO this has issues for components with multiple inputs (on/off paths, thermostat)
    handleInputChange = (id, param, value) => {
        console.log(param)
        this.setState(prevState => {
            prevState[id][param] = value;
            return prevState;
        });
    };

    // Render 2 column layout with device and sensor cards
    renderLayout = () => {
        const deviceEntries = (
            <TransitionGroup className="device-container">
                {Object.entries(this.state)
                .filter(([key, _]) => key.startsWith('device'))
                .map(([key, config]) => (
                    <CSSTransition key={key} timeout={500} classNames="fade">
                        <InstanceCard
                            key={key}
                            id={key}
                            category="device"
                            config={config}
                            onDelete={() => this.deleteInstance(key)}
                            onInputChange={(paramName, value) => this.handleInputChange(key, paramName, value)}
                            onTypeChange={(event) => this.changeInstanceType(key, "device", event)}
                        />
                    </CSSTransition>
                ))}
            </TransitionGroup>
        );

        const sensorEntries = (
            <TransitionGroup className="sensor-container">
                {Object.entries(this.state)
                .filter(([key, _]) => key.startsWith('sensor'))
                .map(([key, config]) => (
                    <CSSTransition key={key} timeout={500} classNames="fade">
                        <InstanceCard
                            key={key}
                            id={key}
                            category="sensor"
                            config={config}
                            onDelete={() => this.deleteInstance(key)}
                            onInputChange={(paramName, value) => this.handleInputChange(key, paramName, value)}
                            onTypeChange={(event) => this.changeInstanceType(key, "sensor", event)}
                        />
                    </CSSTransition>
                ))}
            </TransitionGroup>
        );

        return (
            <>
                <h1 className="text-center pt-3 pb-4">TEST</h1>

                <div id="page1" className="d-flex flex-column h-100">
                    <div className="row mt-3">
                        <div id="sensors" className="col-sm">
                            <h2 className="text-center">Add Sensors</h2>
                            {sensorEntries}
                            <div id="add_sensor" className="text-center position-relative">
                                <button className="btn-secondary btn mb-3" onClick={() => this.addInstance('sensor')}>Add Sensor</button>
                            </div>
                        </div>
                        <div id="devices" className="col-sm">
                            <h2 className="text-center">Add Devices</h2>
                            {deviceEntries}
                            <div id="add_device" className="text-center position-relative">
                                <button className="btn-secondary btn mb-3" onClick={() => this.addInstance('device')}>Add Device</button>
                            </div>
                        </div>
                    </div>
                </div>

                <div className="bottom">
                    <button className="log-button" onClick={() => this.logState()}>Log State</button>
                </div>
            </>
        );
    }

    render() {
        return this.renderLayout();
    }
}


// Called by deleteInstance, decrements IDs of all subsequent instances to prevent gaps
// Example: If device2 is deleted, device3 becomes device2, device4 becomes device3, etc
function update_ids(target, state) {
    // Get category (device or sensor) and index of removed instance
    const category = target.replace(/[0-9]/g, '');
    const index = target.replace(/[a-zA-Z]/g, '');

    // Get list of all instances in same category
    var instances = filterObject(state, category);

    // If target is device get list of sensors (used to update target IDs)
    if (category === 'device') {
        var sensors = filterObject(state, 'sensor');
        // Remove device from all sensor targets
        for (const sensor in sensors) {
            state[sensor]['targets'] = state[sensor]['targets'].filter(item => item !== target);
        };
    };

    // Iterate all instances in category starting from the removed instance index
    for (let i=parseInt(index); i<Object.entries(instances).length+1; i++) {
        // Removed index now available, decrement next index by 1
        const new_id = `${category}${i}`;
        const old_id = `${category}${i+1}`;
        state[new_id] = JSON.parse(JSON.stringify(state[old_id]));
        delete state[old_id];

        // Decrement device index in sensor targets lists to match above
        if (category === 'device') {
            for (const sensor in sensors) {
                if (state[sensor]['targets'].includes(old_id)) {
                    state[sensor]['targets'] = state[sensor]['targets'].filter(item => item !== old_id);
                    state[sensor]['targets'].push(new_id);
                };
            };
        };
    }

    return state;
};


export default App;
