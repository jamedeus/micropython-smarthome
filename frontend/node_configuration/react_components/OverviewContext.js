import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';


export const OverviewContext = createContext();

export const OverviewContextProvider = ({ children }) => {
    // Load context set by django template
    const [context, setContext] = useState(() => {
        function parse_dom_context(name) {
            const element = document.getElementById(name);
            return JSON.parse(element.textContent);
        }

        // Parse context elements created by django template
        return {
            not_uploaded: parse_dom_context("not_uploaded"),
            uploaded: parse_dom_context("uploaded"),
            schedule_keywords: parse_dom_context("schedule_keywords"),
            client_ip: parse_dom_context("client_ip")
        };
    });

    const addScheduleKeyword = (keyword, timestamp) => {
        let update = { ...context.schedule_keywords };
        update[keyword] = timestamp;
        setContext({ ...context, ["schedule_keywords"]: update})
    };

    const editScheduleKeyword = (keyword_old, keyword_new, timestamp_new) => {
        let update = { ...context.schedule_keywords };
        if (keyword_old === keyword_new) {
            update[keyword_old] = timestamp_new;
        } else {
            delete update[keyword_old];
            update[keyword_new] = timestamp_new;
        }
        setContext({ ...context, ["schedule_keywords"]: update})
    };

    const deleteScheduleKeyword = (keyword) => {
        let update = { ...context.schedule_keywords };
        delete update[keyword];
        setContext({ ...context, ["schedule_keywords"]: update})
    };

    return (
        <OverviewContext.Provider value={{
            context,
            setContext,
            addScheduleKeyword,
            editScheduleKeyword,
            deleteScheduleKeyword
        }}>
            {children}
        </OverviewContext.Provider>
    );
};

OverviewContextProvider.propTypes = {
    children: PropTypes.node,
};
