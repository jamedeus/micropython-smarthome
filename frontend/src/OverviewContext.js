import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';
import { parse_dom_context } from 'util/django_util';


export const OverviewContext = createContext();

export const OverviewContextProvider = ({ children }) => {
    // Load context set by django template
    const [context, setContext] = useState(() => {
        return {
            not_uploaded: parse_dom_context("not_uploaded"),
            uploaded: parse_dom_context("uploaded")
        };
    });

    // Adds node to ExistingNodesTable (only used by RestoreModal)
    const addNewNode = (friendly_name, filename, ip) => {
        setContext({ ...context, uploaded: [
            ...context.uploaded,
            {friendly_name: friendly_name, filename: filename, ip: ip}
        ]});
    };

    // Deletes node from ExistingNodesTable
    const deleteExistingNode = (friendly_name) => {
        setContext({ ...context, uploaded: context.uploaded.filter(
            node => node.friendly_name !== friendly_name
        )});
    };

    // Changes IP of node on ExistingNodesTable
    const changeExistingNodeIp = (friendly_name, newIp) => {
        setContext({ ...context, uploaded: context.uploaded.map(node =>
            node.friendly_name === friendly_name ? { ...node, ip: newIp} : node
        )});
    };

    // Delete config from NewConfigTable
    const deleteNewConfig = (filename) => {
        setContext({ ...context, not_uploaded: context.not_uploaded.filter(
            config => config.filename !== filename
        )});
    };

    // Adds new node to ExistingNodesTable, removes config from NewConfigTable
    const handleNewConfigUpload = (friendly_name, filename, ip) => {
        const uploaded = [
            ...context.uploaded,
            {friendly_name: friendly_name, filename: filename, ip: ip}
        ];
        const not_uploaded = context.not_uploaded.filter(
            config => config.filename !== filename
        );
        setContext({ ...context,
            uploaded: uploaded,
            not_uploaded: not_uploaded
        });
    };

    return (
        <OverviewContext.Provider value={{
            context,
            setContext,
            addNewNode,
            deleteExistingNode,
            changeExistingNodeIp,
            deleteNewConfig,
            handleNewConfigUpload
        }}>
            {children}
        </OverviewContext.Provider>
    );
};

OverviewContextProvider.propTypes = {
    children: PropTypes.node,
};
