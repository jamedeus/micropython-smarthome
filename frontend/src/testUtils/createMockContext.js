// Creates DOM elements containing JSON (mocks django json_script filter)
// Takes context id (arg passed to parse_dom_context) and json contents
const createMockContext = (id, json) => {
    const mockContext = document.createElement('div');
    mockContext.id = id;
    mockContext.textContent = JSON.stringify(json);
    document.body.appendChild(mockContext);
};

export default createMockContext;
