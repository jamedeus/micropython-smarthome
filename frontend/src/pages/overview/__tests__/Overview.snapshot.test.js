import React from 'react';
import createMockContext from 'src/testUtils/createMockContext';
import App from '../Overview';
import { OverviewContextProvider } from 'root/OverviewContext';
import { mockContext } from './mockContext';

describe('App', () => {
    it('matches snapshot with new configs and existing nodes', () => {
        // Create mock state objects
        createMockContext('not_uploaded', mockContext.not_uploaded);
        createMockContext('uploaded', mockContext.uploaded);
        createMockContext('schedule_keywords', mockContext.schedule_keywords);
        createMockContext('desktop_integration_link', mockContext.desktop_integration_link);
        createMockContext('client_ip', mockContext.client_ip);

        // Render App, confirm matches snapshot
        const component = render(
            <OverviewContextProvider>
                <App />
            </OverviewContextProvider>
        );
        expect(component).toMatchSnapshot();
    });
});
