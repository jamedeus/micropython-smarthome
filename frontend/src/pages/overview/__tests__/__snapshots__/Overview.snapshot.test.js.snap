// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot with new configs and existing nodes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="not_uploaded"
    >
      [{"filename":"new-config.json","friendly_name":"new config"}]
    </div>
    <div
      id="uploaded"
    >
      [{"friendly_name":"Bathroom","ip":"192.168.1.100","filename":"bathroom.json"},{"friendly_name":"Kitchen","ip":"192.168.1.101","filename":"kitchen.json"},{"friendly_name":"Living Room","ip":"192.168.1.102","filename":"living-room.json"},{"friendly_name":"Bedroom","ip":"192.168.1.103","filename":"bedroom.json"}]
    </div>
    <div
      id="schedule_keywords"
    >
      [{"id":2,"keyword":"morning","timestamp":"08:00"},{"id":3,"keyword":"sleep","timestamp":"23:00"},{"id":4,"keyword":"relax","timestamp":"20:00"}]
    </div>
    <div
      id="desktop_integration_link"
    >
      "/static/node_configuration/micropython-smarthome-integration.zip"
    </div>
    <div
      id="client_ip"
    >
      "127.0.0.1"
    </div>
    <div>
      <div
        class="d-flex flex-column vh-100"
      >
        <div
          class="d-flex justify-content-between"
        >
          <button
            class="btn btn-primary"
            style="visibility: hidden;"
            type="button"
          >
            <i
              class="bi-chevron-left"
            />
          </button>
          <h1
            class="my-3"
          >
            Configure Nodes
          </h1>
          <div
            class="my-auto dropdown"
          >
            <button
              aria-expanded="false"
              class="dropdown-toggle btn btn-light"
              id="settings-button"
              type="button"
            >
              <i
                class="bi-gear-fill"
              />
            </button>
          </div>
        </div>
        <div
          class="text-center section pt-2 px-0 mb-5 row"
          id="not_uploaded"
        >
          <h3
            class="text-center my-1"
          >
            Configs Ready to Upload
          </h3>
          <div
            class="collapse show"
          >
            <table
              class="mt-3 mx-auto table table-sm table-borderless table-hover"
            >
              <thead>
                <tr>
                  <th
                    class="w-50"
                  >
                    Name
                  </th>
                  <th
                    class="w-50"
                  >
                    IP
                  </th>
                  <th />
                  <th />
                </tr>
              </thead>
              <tbody>
                <tr
                  id="new-config.json"
                >
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control text-center"
                    >
                      new-config.json
                    </span>
                  </td>
                  <td
                    class="align-middle"
                  >
                    <input
                      class="text-center ip-input form-control"
                      id="new-config.json-ip"
                      placeholder="xxx.xxx.x.xxx"
                      type="text"
                      value=""
                    />
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <button
                      class="btn btn-primary btn-sm"
                      disabled=""
                      type="button"
                    >
                      Upload
                    </button>
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <button
                      class="btn btn-danger btn-sm"
                      type="button"
                    >
                      <i
                        class="bi-trash"
                      />
                    </button>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          class="text-center section px-0 pt-2 row"
          id="existing"
        >
          <h3
            class="text-center my-1"
          >
            Existing Nodes
          </h3>
          <div
            class="collapse show"
          >
            <table
              class="table-borderless table-sm table-hover mt-3 mx-auto table"
            >
              <thead>
                <tr>
                  <th
                    class="w-50"
                  >
                    Name
                  </th>
                  <th
                    class="w-50"
                  >
                    IP
                  </th>
                  <th />
                </tr>
              </thead>
              <tbody>
                <tr
                  id="Bathroom"
                >
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      Bathroom
                    </span>
                  </td>
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      192.168.1.100
                    </span>
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <div
                      class="my-auto dropdown"
                    >
                      <button
                        aria-expanded="false"
                        class="dropdown-toggle btn btn-primary btn-sm"
                        id="react-aria-:r1:"
                        type="button"
                      >
                        <i
                          class="bi-list"
                        />
                      </button>
                    </div>
                  </td>
                </tr>
                <tr
                  id="Kitchen"
                >
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      Kitchen
                    </span>
                  </td>
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      192.168.1.101
                    </span>
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <div
                      class="my-auto dropdown"
                    >
                      <button
                        aria-expanded="false"
                        class="dropdown-toggle btn btn-primary btn-sm"
                        id="react-aria-:r2:"
                        type="button"
                      >
                        <i
                          class="bi-list"
                        />
                      </button>
                    </div>
                  </td>
                </tr>
                <tr
                  id="Living Room"
                >
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      Living Room
                    </span>
                  </td>
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      192.168.1.102
                    </span>
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <div
                      class="my-auto dropdown"
                    >
                      <button
                        aria-expanded="false"
                        class="dropdown-toggle btn btn-primary btn-sm"
                        id="react-aria-:r3:"
                        type="button"
                      >
                        <i
                          class="bi-list"
                        />
                      </button>
                    </div>
                  </td>
                </tr>
                <tr
                  id="Bedroom"
                >
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      Bedroom
                    </span>
                  </td>
                  <td
                    class="align-middle"
                  >
                    <span
                      class="form-control keyword text-center"
                    >
                      192.168.1.103
                    </span>
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <div
                      class="my-auto dropdown"
                    >
                      <button
                        aria-expanded="false"
                        class="dropdown-toggle btn btn-primary btn-sm"
                        id="react-aria-:r4:"
                        type="button"
                      >
                        <i
                          class="bi-list"
                        />
                      </button>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          class="mt-2 mb-5 text-center"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Create new config
          </button>
        </div>
        <div
          class="text-center section px-0 pt-2 row"
          id="keywords"
        >
          <h3
            class="text-center my-1"
          >
            Schedule Keywords
          </h3>
          <div
            class="collapse show"
          >
            <table
              class="table-borderless table-sm table-hover mt-3 mx-auto table"
            >
              <thead>
                <tr>
                  <th
                    class="w-50"
                  >
                    Keyword
                  </th>
                  <th
                    class="w-50"
                  >
                    Timestamp
                  </th>
                  <th />
                </tr>
              </thead>
              <tbody>
                <tr
                  id="morning_row"
                >
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      placeholder="Keyword"
                      type="text"
                      value="morning"
                    />
                  </td>
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      type="time"
                      value="08:00"
                    />
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <button
                      class="btn btn-danger btn-sm"
                      type="button"
                    >
                      <i
                        class="bi-trash"
                      />
                    </button>
                  </td>
                </tr>
                <tr
                  id="sleep_row"
                >
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      placeholder="Keyword"
                      type="text"
                      value="sleep"
                    />
                  </td>
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      type="time"
                      value="23:00"
                    />
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <button
                      class="btn btn-danger btn-sm"
                      type="button"
                    >
                      <i
                        class="bi-trash"
                      />
                    </button>
                  </td>
                </tr>
                <tr
                  id="relax_row"
                >
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      placeholder="Keyword"
                      type="text"
                      value="relax"
                    />
                  </td>
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      type="time"
                      value="20:00"
                    />
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <button
                      class="btn btn-danger btn-sm"
                      type="button"
                    >
                      <i
                        class="bi-trash"
                      />
                    </button>
                  </td>
                </tr>
                <tr
                  id="_row"
                >
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      placeholder="Keyword"
                      type="text"
                      value=""
                    />
                  </td>
                  <td
                    class="align-middle"
                  >
                    <input
                      class="keyword text-center form-control"
                      type="time"
                      value=""
                    />
                  </td>
                  <td
                    class="min align-middle"
                  >
                    <button
                      class="btn btn-primary btn-sm"
                      disabled=""
                      type="button"
                    >
                      <i
                        class="bi-plus"
                      />
                    </button>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          class="d-flex align-items-center flex-column mt-auto py-4"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            Frontend
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex flex-column vh-100"
    >
      <div
        class="d-flex justify-content-between"
      >
        <button
          class="btn btn-primary"
          style="visibility: hidden;"
          type="button"
        >
          <i
            class="bi-chevron-left"
          />
        </button>
        <h1
          class="my-3"
        >
          Configure Nodes
        </h1>
        <div
          class="my-auto dropdown"
        >
          <button
            aria-expanded="false"
            class="dropdown-toggle btn btn-light"
            id="settings-button"
            type="button"
          >
            <i
              class="bi-gear-fill"
            />
          </button>
        </div>
      </div>
      <div
        class="text-center section pt-2 px-0 mb-5 row"
        id="not_uploaded"
      >
        <h3
          class="text-center my-1"
        >
          Configs Ready to Upload
        </h3>
        <div
          class="collapse show"
        >
          <table
            class="mt-3 mx-auto table table-sm table-borderless table-hover"
          >
            <thead>
              <tr>
                <th
                  class="w-50"
                >
                  Name
                </th>
                <th
                  class="w-50"
                >
                  IP
                </th>
                <th />
                <th />
              </tr>
            </thead>
            <tbody>
              <tr
                id="new-config.json"
              >
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control text-center"
                  >
                    new-config.json
                  </span>
                </td>
                <td
                  class="align-middle"
                >
                  <input
                    class="text-center ip-input form-control"
                    id="new-config.json-ip"
                    placeholder="xxx.xxx.x.xxx"
                    type="text"
                    value=""
                  />
                </td>
                <td
                  class="min align-middle"
                >
                  <button
                    class="btn btn-primary btn-sm"
                    disabled=""
                    type="button"
                  >
                    Upload
                  </button>
                </td>
                <td
                  class="min align-middle"
                >
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                    <i
                      class="bi-trash"
                    />
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="text-center section px-0 pt-2 row"
        id="existing"
      >
        <h3
          class="text-center my-1"
        >
          Existing Nodes
        </h3>
        <div
          class="collapse show"
        >
          <table
            class="table-borderless table-sm table-hover mt-3 mx-auto table"
          >
            <thead>
              <tr>
                <th
                  class="w-50"
                >
                  Name
                </th>
                <th
                  class="w-50"
                >
                  IP
                </th>
                <th />
              </tr>
            </thead>
            <tbody>
              <tr
                id="Bathroom"
              >
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    Bathroom
                  </span>
                </td>
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    192.168.1.100
                  </span>
                </td>
                <td
                  class="min align-middle"
                >
                  <div
                    class="my-auto dropdown"
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle btn btn-primary btn-sm"
                      id="react-aria-:r1:"
                      type="button"
                    >
                      <i
                        class="bi-list"
                      />
                    </button>
                  </div>
                </td>
              </tr>
              <tr
                id="Kitchen"
              >
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    Kitchen
                  </span>
                </td>
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    192.168.1.101
                  </span>
                </td>
                <td
                  class="min align-middle"
                >
                  <div
                    class="my-auto dropdown"
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle btn btn-primary btn-sm"
                      id="react-aria-:r2:"
                      type="button"
                    >
                      <i
                        class="bi-list"
                      />
                    </button>
                  </div>
                </td>
              </tr>
              <tr
                id="Living Room"
              >
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    Living Room
                  </span>
                </td>
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    192.168.1.102
                  </span>
                </td>
                <td
                  class="min align-middle"
                >
                  <div
                    class="my-auto dropdown"
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle btn btn-primary btn-sm"
                      id="react-aria-:r3:"
                      type="button"
                    >
                      <i
                        class="bi-list"
                      />
                    </button>
                  </div>
                </td>
              </tr>
              <tr
                id="Bedroom"
              >
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    Bedroom
                  </span>
                </td>
                <td
                  class="align-middle"
                >
                  <span
                    class="form-control keyword text-center"
                  >
                    192.168.1.103
                  </span>
                </td>
                <td
                  class="min align-middle"
                >
                  <div
                    class="my-auto dropdown"
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle btn btn-primary btn-sm"
                      id="react-aria-:r4:"
                      type="button"
                    >
                      <i
                        class="bi-list"
                      />
                    </button>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="mt-2 mb-5 text-center"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Create new config
        </button>
      </div>
      <div
        class="text-center section px-0 pt-2 row"
        id="keywords"
      >
        <h3
          class="text-center my-1"
        >
          Schedule Keywords
        </h3>
        <div
          class="collapse show"
        >
          <table
            class="table-borderless table-sm table-hover mt-3 mx-auto table"
          >
            <thead>
              <tr>
                <th
                  class="w-50"
                >
                  Keyword
                </th>
                <th
                  class="w-50"
                >
                  Timestamp
                </th>
                <th />
              </tr>
            </thead>
            <tbody>
              <tr
                id="morning_row"
              >
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    placeholder="Keyword"
                    type="text"
                    value="morning"
                  />
                </td>
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    type="time"
                    value="08:00"
                  />
                </td>
                <td
                  class="min align-middle"
                >
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                    <i
                      class="bi-trash"
                    />
                  </button>
                </td>
              </tr>
              <tr
                id="sleep_row"
              >
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    placeholder="Keyword"
                    type="text"
                    value="sleep"
                  />
                </td>
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    type="time"
                    value="23:00"
                  />
                </td>
                <td
                  class="min align-middle"
                >
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                    <i
                      class="bi-trash"
                    />
                  </button>
                </td>
              </tr>
              <tr
                id="relax_row"
              >
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    placeholder="Keyword"
                    type="text"
                    value="relax"
                  />
                </td>
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    type="time"
                    value="20:00"
                  />
                </td>
                <td
                  class="min align-middle"
                >
                  <button
                    class="btn btn-danger btn-sm"
                    type="button"
                  >
                    <i
                      class="bi-trash"
                    />
                  </button>
                </td>
              </tr>
              <tr
                id="_row"
              >
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    placeholder="Keyword"
                    type="text"
                    value=""
                  />
                </td>
                <td
                  class="align-middle"
                >
                  <input
                    class="keyword text-center form-control"
                    type="time"
                    value=""
                  />
                </td>
                <td
                  class="min align-middle"
                >
                  <button
                    class="btn btn-primary btn-sm"
                    disabled=""
                    type="button"
                  >
                    <i
                      class="bi-plus"
                    />
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="d-flex align-items-center flex-column mt-auto py-4"
      >
        <button
          class="btn btn-secondary"
          type="button"
        >
          Frontend
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
